balancer.go

```diff
@@ -288,3 +288,82 @@ func murmur2(data []byte) uint32 {
 
 	return h
 }
+
+// StickyPartitioner is an Balancer implementation that the Message key is NOT used as part of the balancing strategy
+// in this balancer. Messages with the same key are not guaranteed to be sent to the same partition.
+// If a partition is specified in the Message, use it
+// Otherwise choose the sticky partition that changes when the batch is full.
+type StickyPartitioner struct {
+	spc stickyPartitionCache
+}
+
+// Balance satisfies the Balancer interface.
+func (sp *StickyPartitioner) Balance(msg Message, partitions ...int) int {
+	return sp.balance(msg, partitions)
+}
+
+func (sp *StickyPartitioner) balance(msg Message, partitions []int) int {
+	return sp.spc.partition(msg, partitions)
+}
+
+// OnNewBatch changes the sticky partition If a batch completed for the current sticky partition.
+// Alternately, if no sticky partition has been determined, set one.
+func (sp *StickyPartitioner) OnNewBatch(msg Message, partitions []int, prevPartition int) int {
+	return sp.spc.nextPartition(msg, partitions, prevPartition)
+}
+
+// stickyPartitionCache implements a cache used for sticky partitioning behavior. The cache tracks the current sticky
+// partition for any given topic.
+type stickyPartitionCache struct {
+	lock       sync.Mutex
+	indexCache map[string]int
+}
+
+func (spc *stickyPartitionCache) nextPartition(msg Message, partitions []int, prevPartition int) int {
```

Comment:
I'm not quite sure I understand the purpose of the `prevPartition` argument, I think the code might work without it

The way I understand it the logic for selecting the next partition is:
- If there is only 1 partition return that partition
- otherwise, pick a random partition, save it, and return that